api/include/opentelemetry/trace/trace_state.h

```diff
@@ -139,8 +269,85 @@ class TraceState
     return nostd::span<Entry>(entries_.get(), num_entries_);
   }
 
-  // Returns whether key is a valid key. See https://www.w3.org/TR/trace-context/#key
+  /** Returns whether key is a valid key. See https://www.w3.org/TR/trace-context/#key
+   * Identifiers MUST begin with a lowercase letter or a digit, and can only contain
+   * lowercase letters (a-z), digits (0-9), underscores (_), dashes (-), asterisks (*),
+   * and forward slashes (/).
+   * For multi-tenant vendor scenarios, an at sign (@) can be used to prefix the vendor name.
+   *
+   */
   static bool IsValidKey(nostd::string_view key)
+  {
+#if HAVE_WORKING_REGEX
+    return IsValidKeyRegEx(key);
+#else
+    return IsValidKeyNonRegEx(key);
+#endif
+  }
+
+  /** Returns whether value is a valid value. See https://www.w3.org/TR/trace-context/#value
+   * The value is an opaque string containing up to 256 printable ASCII (RFC0020)
+   *  characters ((i.e., the range 0x20 to 0x7E) except comma , and equal =)
+   */
+  static bool IsValidValue(nostd::string_view value)
+  {
+#if HAVE_WORKING_REGEX
+    return IsValidValueRegEx(value);
+#else
+    return IsValidValueNonRegEx(value);
+#endif
+  }
+
+private:
+  // Store entries in a C-style array to avoid using std::array or std::vector.
+  nostd::unique_ptr<Entry[]> entries_;
+
+  // Maintain the number of entries in entries_. Must be in the range [0, kMaxKeyValuePairs].
+  size_t num_entries_;
+
+  // An empty TraceState.
+  TraceState() noexcept : entries_(new Entry[kMaxKeyValuePairs]), num_entries_(0) {}
+
+  static nostd::string_view TrimString(nostd::string_view str, size_t left, size_t right)
+  {
+    while (str[(std::size_t)right] == ' ' && left < right)
+    {
+      right--;
+    }
+    while (str[(std::size_t)left] == ' ' && left < right)
+    {
+      left++;
+    }
+    return str.substr(left, right - left + 1);
+  }
+
+  static bool IsValidKeyRegEx(nostd::string_view key)
+  {
+    std::regex reg_key("^[a-z0-9][a-z0-9*_\\-/]{0,255}$");
+    std::regex reg_key_multitenant(
+        "^[a-z0-9][a-z0-9*_\\-/]{0,240}(@)[a-z0-9][a-z0-9*_\\-/]{0,13}$");
+    std::string key_s(key);
+    if (std::regex_match(key_s, reg_key) || std::regex_match(key_s, reg_key_multitenant))
+    {
+      return true;
+    }
+    return false;
+  }
+
+  static bool IsValidValueRegEx(nostd::string_view value)
+  {
+    // Hex 0x20 to 0x2B, 0x2D to 0x3C, 0x3E to 0x7E
+    std::regex reg_value(
+        "^[\\x20-\\x2B\\x2D-\\x3C\\x3E-\\x7E]{0,255}[\\x21-\\x2B\\x2D-\\x3C\\x3E-\\x7E]$");
+    // Need to benchmark without regex, as a string object is created here.
+    if (std::regex_match(std::string(value), reg_value))
+    {
+      return true;
+    }
+    return false;
```

Comment:
```suggestion
    return std::regex_match(std::string(value), reg_value);
```
But that's more a matter of taste.

Reply:
This is definitely better :)